### DISCUSSION TITLE
ESP32 I2S 24 Parallel channel support

### DISCUSSION THREAD

**Discussion by Makuna** [Makuna/NeoPixelBus#538] [Announcements]
This issue (https://github.com/Makuna/NeoPixelBus/issues/270) has been around for a while but I haven't been motivated enough to tackle it.   

I need some incentive.  As I write this, this repo currently has 935 Stars, so if we can reach 1000 Stars in the next few weeks, this will motivate me to jump on this task and get it done for the new year or sooner.  

There are also the monetary incentive methods for those who have already starred the repo listed in the wiki and readme; but I like the gamification starring brings to this conversation.

**Comment by plaber** [comment]
Congratulations! You riched 1k. :-) I hope you add this futures soon. Thanks!

**Reply by Makuna** [reply]
I am in the debugging phase now.

**Comment by kleurbleur** [comment]
After reading both issue https://github.com/Makuna/NeoPixelBus/issues/270 and above, do I understand that currently NeoPixelBus does not support multiple data pin outputs?

For my project I want to translate artnet universes and distribute them over predefined output pins of an ESP32.

**Reply by Makuna** [reply]
I2s parallel mode as mentioned in this discussion and the linked issue you point at are different topic than multi-channel support in general.  The i2s parallel is a specific advanced method to achieve multi-channel. 

ESP32 supports today, up to 8 channels of RMT and two channels of i2s depending on the model of the ESP32.  The newer models often only support 4 channels of RMT and one i2s.  Even with 8 channels of RMT, after four there is a delay issues in the core of when they are triggered versus the first four.  See https://github.com/Makuna/NeoPixelBus/wiki/ESP32-NeoMethods for more details.  Also search through discussions for NeoPixelBus_Channel to understand indexed bus support which isn't documented yet in the wiki.

**Comment by plaber** [comment]
Plz, say is any progress of this function?

**Reply by Makuna** [reply]
I have it written; it just doesn't output anything on the pins.  Not sure why as I followed through some other source and the only thing I can deduce is that that the 8bit parallel mode doesn't work at all.  The other libraries seem to use higher bit parallel (16/24) and waist memory for the bits unused.  
So it's on hold until I can get some guidance or inspiration.

**Reply by awawa-dev** [reply]
@Makuna  Is there any chance to preview the changes? I am willing to investigate the issue.

**Reply by Makuna** [reply]
@awawa-dev I published the active development branch for the work.  See [Esp32i2sX branch](https://github.com/Makuna/NeoPixelBus/tree/Esp32I2sX).

There is an example sketch I currently use for testing under examples/esp32/NeoPixel_ESP32_I2sParallel.ino

**Reply by awawa-dev** [reply]
Thank you 👍 I'll try to help.

**Reply by Makuna** [reply]
Thanks to @awawa-dev the blocking issues have been resolved and I am cleaning up the code and testing.  Still need to cleanup normal single channel and test it also before this gets merged into master.
Focusing on 8bit mode first, but 16 bit mode was tested by awawa-dev.
see [Esp32i2sX branch](https://github.com/Makuna/NeoPixelBus/tree/Esp32I2sX) if you want to live on bleeding edge.  There is an example under the examples\esp32 that demonstrate it but only uses 4 of 8.

I just demonstrated 8 400 pixel strips (3200 pixels!) being updated as fast as a sketch can run (dirty, show, loop) and I am seeing 80fps and solid signals.  Now of course, the real speed will depend on the users update code, but it proved the async nature of the update.

**Comment by Makuna** [comment]
A big question, what IO pins would you use if you have 8 channels available to you?  16 channels?  24 channels?

**Comment by Makuna** [comment]
https://github.com/Makuna/NeoPixelBus/tree/Esp32I2sX
Will be merging this into the main branch within the week.
Supports ESP32 and ESP32S2 (C3 and S3 can't use i2s for parallel mode, they have a new LED peripheral).
Supports x8 and x16 channel modes.  I didn't even try 24 bit mode but I am unsure more than 16 will ever be used and testing matrix was already taken a long time due to S2 being different let along i2s0 and i2s1 on ESP32 being different.

**Comment by Makuna** [comment]
Ok, its merged into main.
